{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process bednot() output for RefSeq vs ClinVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "from datetime import datetime\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "today = datetime.today().strftime('%Y-%m-%d')\n",
    "os.chdir('/Users/pbousounis/Experiments/2019-10-29_hg19mod/2019-10-25_RefSeq_GFF3_download_validate')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref2clin_file = '2019-11-05_GRCh37_latest_genomic_IN_2019-11-05_clinvar_path-l_RECOVERY.txt'\n",
    "ref2clin = pd.read_csv(ref2clin_file, sep='\\t', low_memory=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>clinvar_recovery_count</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12720</td>\n",
       "      <td>DDX11L1-NR_046018:exon2(+)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14408</td>\n",
       "      <td>DDX11L1-NR_046018:exon3(+)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>29320</td>\n",
       "      <td>29369</td>\n",
       "      <td>WASH7P-NR_024540:exon11(-)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>24737</td>\n",
       "      <td>24890</td>\n",
       "      <td>WASH7P-NR_024540:exon10(-)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>18267</td>\n",
       "      <td>18365</td>\n",
       "      <td>WASH7P-NR_024540:exon9(-)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr  start    end                     exon_id  clinvar_recovery_count  \\\n",
       "0   1  12612  12720  DDX11L1-NR_046018:exon2(+)                       0   \n",
       "1   1  13220  14408  DDX11L1-NR_046018:exon3(+)                       0   \n",
       "2   1  29320  29369  WASH7P-NR_024540:exon11(-)                       0   \n",
       "3   1  24737  24890  WASH7P-NR_024540:exon10(-)                       0   \n",
       "4   1  18267  18365   WASH7P-NR_024540:exon9(-)                       0   \n",
       "\n",
       "       tx_id     gene  \n",
       "0  NR_046018  DDX11L1  \n",
       "1  NR_046018  DDX11L1  \n",
       "2  NR_024540   WASH7P  \n",
       "3  NR_024540   WASH7P  \n",
       "4  NR_024540   WASH7P  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = ref2clin\n",
    "\n",
    "# extract RefSeq TX accessions from exon_ids\n",
    "df['tx_id'] = df['exon_id'].str.split('-', expand=True)[1].str.split(':', expand=True)[0]\n",
    "df['gene'] = df['exon_id'].str.split('-', expand=True)[0]\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>gene</th>\n",
       "      <th>total_gene_vars</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>949522</td>\n",
       "      <td>949523</td>\n",
       "      <td>183381_ISG15</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>949695</td>\n",
       "      <td>949696</td>\n",
       "      <td>161455_ISG15</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>949738</td>\n",
       "      <td>949739</td>\n",
       "      <td>161454_ISG15</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>957604</td>\n",
       "      <td>957605</td>\n",
       "      <td>243036_AGRN</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>957692</td>\n",
       "      <td>957693</td>\n",
       "      <td>243037_AGRN</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>15.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end          name   gene  total_gene_vars\n",
       "0  chr1  949522  949523  183381_ISG15  ISG15              3.0\n",
       "1  chr1  949695  949696  161455_ISG15  ISG15              3.0\n",
       "2  chr1  949738  949739  161454_ISG15  ISG15              3.0\n",
       "3  chr1  957604  957605   243036_AGRN   AGRN             15.0\n",
       "4  chr1  957692  957693   243037_AGRN   AGRN             15.0"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cv_bed_file = '/Users/pbousounis/Experiments/2019-10-29_hg19mod/2019-11-06_refseq_cv_slop_intersect/bed/2019-11-05_clinvar_path-l.bed'\n",
    "cv_bed = pd.read_csv(cv_bed_file, sep='\\t')\n",
    "cv_bed.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Join the clinvar total_gene_vars to the ref2clin table, by gene names\n",
    "\n",
    "* Add transcript recovery and percent recovery columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_x</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>clinvar_recovery_count</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_recovery</th>\n",
       "      <th>transcript_recovery</th>\n",
       "      <th>tx_perc_recovery</th>\n",
       "      <th>chr_y</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>name</th>\n",
       "      <th>total_gene_vars</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>12612</td>\n",
       "      <td>12720</td>\n",
       "      <td>DDX11L1-NR_046018:exon2(+)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>13220</td>\n",
       "      <td>14408</td>\n",
       "      <td>DDX11L1-NR_046018:exon3(+)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_046018</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>29320</td>\n",
       "      <td>29369</td>\n",
       "      <td>WASH7P-NR_024540:exon11(-)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>24737</td>\n",
       "      <td>24890</td>\n",
       "      <td>WASH7P-NR_024540:exon10(-)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>18267</td>\n",
       "      <td>18365</td>\n",
       "      <td>WASH7P-NR_024540:exon9(-)</td>\n",
       "      <td>0</td>\n",
       "      <td>NR_024540</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr_x  start_x  end_x                     exon_id  clinvar_recovery_count  \\\n",
       "0     1    12612  12720  DDX11L1-NR_046018:exon2(+)                       0   \n",
       "1     1    13220  14408  DDX11L1-NR_046018:exon3(+)                       0   \n",
       "2     1    29320  29369  WASH7P-NR_024540:exon11(-)                       0   \n",
       "3     1    24737  24890  WASH7P-NR_024540:exon10(-)                       0   \n",
       "4     1    18267  18365   WASH7P-NR_024540:exon9(-)                       0   \n",
       "\n",
       "       tx_id     gene  gene_recovery  transcript_recovery  tx_perc_recovery  \\\n",
       "0  NR_046018  DDX11L1            0.0                  0.0               NaN   \n",
       "1  NR_046018  DDX11L1            0.0                  0.0               NaN   \n",
       "2  NR_024540   WASH7P            0.0                  0.0               NaN   \n",
       "3  NR_024540   WASH7P            0.0                  0.0               NaN   \n",
       "4  NR_024540   WASH7P            0.0                  0.0               NaN   \n",
       "\n",
       "  chr_y  start_y  end_y name  total_gene_vars  \n",
       "0   NaN      NaN    NaN  NaN              NaN  \n",
       "1   NaN      NaN    NaN  NaN              NaN  \n",
       "2   NaN      NaN    NaN  NaN              NaN  \n",
       "3   NaN      NaN    NaN  NaN              NaN  \n",
       "4   NaN      NaN    NaN  NaN              NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result = pd.merge(df, cv_bed, how='left', on=['gene'])\n",
    "result.head()\n",
    "# df['gene_recovery'] = \n",
    "# #df['gene_recovery'] = df['clinvar_recovery_count'].groupby(df['gene']).transform('sum')\n",
    "# df['transcript_recovery'] = df['clinvar_recovery_count'].groupby(df['tx_id']).transform('sum')\n",
    "# df['tx_perc_recovery'] = df['transcript_recovery'] / df['gene_recovery'] * 100\n",
    "# df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the file as tsv with date prefix\n",
    "file_out = today + '_GRCh37_latest_genomic_VS_clinvar_path-l_tx-gene-recovery.tsv'\n",
    "df.to_csv(file_out, sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import clinvar bed file for total_gene_vars values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_recovery</th>\n",
       "      <th>transcript_recovery</th>\n",
       "      <th>tx_perc_recovery</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>NM_005101</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>NM_198576</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>50.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>NM_003327</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>NM_080605</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tx_id     gene  gene_recovery  transcript_recovery  \\\n",
       "194     NM_005101    ISG15            3.0                  3.0   \n",
       "196  NM_001305275     AGRN           30.0                 15.0   \n",
       "235     NM_198576     AGRN           30.0                 15.0   \n",
       "345     NM_003327  TNFRSF4            1.0                  1.0   \n",
       "366     NM_080605  B3GALT6           18.0                 18.0   \n",
       "\n",
       "     tx_perc_recovery  \n",
       "194             100.0  \n",
       "196              50.0  \n",
       "235              50.0  \n",
       "345             100.0  \n",
       "366             100.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "txdf = df[['tx_id', 'gene', 'gene_recovery', 'transcript_recovery', 'tx_perc_recovery']].dropna().drop_duplicates()\n",
    "txdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_recovery</th>\n",
       "      <th>transcript_recovery</th>\n",
       "      <th>tx_perc_recovery</th>\n",
       "      <th>tx_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>NM_005101</td>\n",
       "      <td>ISG15</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>NM_198576</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>NM_003327</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>NM_080605</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>NM_004421</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>34.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>599</th>\n",
       "      <td>NM_001330311</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>34.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>NM_001282582</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>NM_001282583</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>NM_001282584</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>647</th>\n",
       "      <td>NM_032348</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>657</th>\n",
       "      <td>NM_001282585</td>\n",
       "      <td>MXRA8</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>810</th>\n",
       "      <td>NM_018188</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>826</th>\n",
       "      <td>NM_001170535</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>NM_001170536</td>\n",
       "      <td>ATAD3A</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>858</th>\n",
       "      <td>NM_001114748</td>\n",
       "      <td>TMEM240</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1309</th>\n",
       "      <td>NM_001282539</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>32.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>34.375000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1320</th>\n",
       "      <td>NM_001282538</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>32.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>31.250000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1330</th>\n",
       "      <td>NM_002074</td>\n",
       "      <td>GNB1</td>\n",
       "      <td>32.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>34.375000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>NM_003036</td>\n",
       "      <td>SKI</td>\n",
       "      <td>15.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1572</th>\n",
       "      <td>NM_002617</td>\n",
       "      <td>PEX10</td>\n",
       "      <td>39.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>46.153846</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578</th>\n",
       "      <td>NM_153818</td>\n",
       "      <td>PEX10</td>\n",
       "      <td>39.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>53.846154</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1792</th>\n",
       "      <td>NM_022114</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1809</th>\n",
       "      <td>NM_199454</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2101</th>\n",
       "      <td>NM_014704</td>\n",
       "      <td>CEP104</td>\n",
       "      <td>5.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2215</th>\n",
       "      <td>NM_001291594</td>\n",
       "      <td>NPHP4</td>\n",
       "      <td>86.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>23.255814</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2241</th>\n",
       "      <td>NM_001291593</td>\n",
       "      <td>NPHP4</td>\n",
       "      <td>86.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>25.581395</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2268</th>\n",
       "      <td>NR_111987</td>\n",
       "      <td>NPHP4</td>\n",
       "      <td>86.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>25.581395</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2301</th>\n",
       "      <td>NM_015102</td>\n",
       "      <td>NPHP4</td>\n",
       "      <td>86.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>25.581395</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2505</th>\n",
       "      <td>NM_207370</td>\n",
       "      <td>GPR153</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8043</th>\n",
       "      <td>NM_001195010</td>\n",
       "      <td>GRHL3</td>\n",
       "      <td>35.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>25.714286</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8059</th>\n",
       "      <td>NM_021180</td>\n",
       "      <td>GRHL3</td>\n",
       "      <td>35.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>25.714286</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8433</th>\n",
       "      <td>NM_013943</td>\n",
       "      <td>CLIC4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8476</th>\n",
       "      <td>NR_135788</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8480</th>\n",
       "      <td>NM_001321772</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8485</th>\n",
       "      <td>NR_135787</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8491</th>\n",
       "      <td>NR_135789</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8495</th>\n",
       "      <td>NR_135777</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8499</th>\n",
       "      <td>NM_020317</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8504</th>\n",
       "      <td>NR_135785</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8510</th>\n",
       "      <td>NR_135144</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8515</th>\n",
       "      <td>NR_135783</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8521</th>\n",
       "      <td>NR_135781</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8527</th>\n",
       "      <td>NR_135782</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8533</th>\n",
       "      <td>NR_135778</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8538</th>\n",
       "      <td>NR_135780</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8543</th>\n",
       "      <td>NR_135786</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8548</th>\n",
       "      <td>NR_135784</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8553</th>\n",
       "      <td>NR_135143</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8557</th>\n",
       "      <td>NR_135792</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8559</th>\n",
       "      <td>NR_135793</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8562</th>\n",
       "      <td>NR_135791</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8565</th>\n",
       "      <td>NR_135790</td>\n",
       "      <td>RSRP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>25.000000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8568</th>\n",
       "      <td>NM_001282872</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8578</th>\n",
       "      <td>NM_001282871</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8587</th>\n",
       "      <td>NM_016124</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8597</th>\n",
       "      <td>NM_001282870</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8606</th>\n",
       "      <td>NM_001282869</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8614</th>\n",
       "      <td>NM_001282868</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8622</th>\n",
       "      <td>NM_001127691</td>\n",
       "      <td>RHD</td>\n",
       "      <td>16.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.500000</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             tx_id     gene  gene_recovery  transcript_recovery  \\\n",
       "194      NM_005101    ISG15            3.0                  3.0   \n",
       "196   NM_001305275     AGRN           30.0                 15.0   \n",
       "235      NM_198576     AGRN           30.0                 15.0   \n",
       "345      NM_003327  TNFRSF4            1.0                  1.0   \n",
       "366      NM_080605  B3GALT6           18.0                 18.0   \n",
       "584      NM_004421     DVL1           34.0                 17.0   \n",
       "599   NM_001330311     DVL1           34.0                 17.0   \n",
       "617   NM_001282582    MXRA8            5.0                  1.0   \n",
       "628   NM_001282583    MXRA8            5.0                  1.0   \n",
       "638   NM_001282584    MXRA8            5.0                  1.0   \n",
       "647      NM_032348    MXRA8            5.0                  1.0   \n",
       "657   NM_001282585    MXRA8            5.0                  1.0   \n",
       "810      NM_018188   ATAD3A            4.0                  2.0   \n",
       "826   NM_001170535   ATAD3A            4.0                  1.0   \n",
       "842   NM_001170536   ATAD3A            4.0                  1.0   \n",
       "858   NM_001114748  TMEM240            6.0                  6.0   \n",
       "1309  NM_001282539     GNB1           32.0                 11.0   \n",
       "1320  NM_001282538     GNB1           32.0                 10.0   \n",
       "1330     NM_002074     GNB1           32.0                 11.0   \n",
       "1525     NM_003036      SKI           15.0                 15.0   \n",
       "1572     NM_002617    PEX10           39.0                 18.0   \n",
       "1578     NM_153818    PEX10           39.0                 21.0   \n",
       "1792     NM_022114   PRDM16           14.0                  7.0   \n",
       "1809     NM_199454   PRDM16           14.0                  7.0   \n",
       "2101     NM_014704   CEP104            5.0                  5.0   \n",
       "2215  NM_001291594    NPHP4           86.0                 20.0   \n",
       "2241  NM_001291593    NPHP4           86.0                 22.0   \n",
       "2268     NR_111987    NPHP4           86.0                 22.0   \n",
       "2301     NM_015102    NPHP4           86.0                 22.0   \n",
       "2505     NM_207370   GPR153            1.0                  1.0   \n",
       "...            ...      ...            ...                  ...   \n",
       "8043  NM_001195010    GRHL3           35.0                  9.0   \n",
       "8059     NM_021180    GRHL3           35.0                  9.0   \n",
       "8433     NM_013943    CLIC4            1.0                  1.0   \n",
       "8476     NR_135788    RSRP1            4.0                  0.0   \n",
       "8480  NM_001321772    RSRP1            4.0                  0.0   \n",
       "8485     NR_135787    RSRP1            4.0                  0.0   \n",
       "8491     NR_135789    RSRP1            4.0                  0.0   \n",
       "8495     NR_135777    RSRP1            4.0                  0.0   \n",
       "8499     NM_020317    RSRP1            4.0                  0.0   \n",
       "8504     NR_135785    RSRP1            4.0                  0.0   \n",
       "8510     NR_135144    RSRP1            4.0                  0.0   \n",
       "8515     NR_135783    RSRP1            4.0                  0.0   \n",
       "8521     NR_135781    RSRP1            4.0                  0.0   \n",
       "8527     NR_135782    RSRP1            4.0                  0.0   \n",
       "8533     NR_135778    RSRP1            4.0                  0.0   \n",
       "8538     NR_135780    RSRP1            4.0                  0.0   \n",
       "8543     NR_135786    RSRP1            4.0                  0.0   \n",
       "8548     NR_135784    RSRP1            4.0                  0.0   \n",
       "8553     NR_135143    RSRP1            4.0                  0.0   \n",
       "8557     NR_135792    RSRP1            4.0                  1.0   \n",
       "8559     NR_135793    RSRP1            4.0                  1.0   \n",
       "8562     NR_135791    RSRP1            4.0                  1.0   \n",
       "8565     NR_135790    RSRP1            4.0                  1.0   \n",
       "8568  NM_001282872      RHD           16.0                  2.0   \n",
       "8578  NM_001282871      RHD           16.0                  2.0   \n",
       "8587     NM_016124      RHD           16.0                  2.0   \n",
       "8597  NM_001282870      RHD           16.0                  2.0   \n",
       "8606  NM_001282869      RHD           16.0                  2.0   \n",
       "8614  NM_001282868      RHD           16.0                  2.0   \n",
       "8622  NM_001127691      RHD           16.0                  2.0   \n",
       "\n",
       "      tx_perc_recovery  tx_rank  \n",
       "194         100.000000      1.0  \n",
       "196          50.000000      1.0  \n",
       "235          50.000000      1.0  \n",
       "345         100.000000      1.0  \n",
       "366         100.000000      1.0  \n",
       "584          50.000000      1.0  \n",
       "599          50.000000      1.0  \n",
       "617          20.000000      1.0  \n",
       "628          20.000000      1.0  \n",
       "638          20.000000      1.0  \n",
       "647          20.000000      1.0  \n",
       "657          20.000000      1.0  \n",
       "810          50.000000      1.0  \n",
       "826          25.000000      2.0  \n",
       "842          25.000000      2.0  \n",
       "858         100.000000      1.0  \n",
       "1309         34.375000      1.0  \n",
       "1320         31.250000      2.0  \n",
       "1330         34.375000      1.0  \n",
       "1525        100.000000      1.0  \n",
       "1572         46.153846      2.0  \n",
       "1578         53.846154      1.0  \n",
       "1792         50.000000      1.0  \n",
       "1809         50.000000      1.0  \n",
       "2101        100.000000      1.0  \n",
       "2215         23.255814      2.0  \n",
       "2241         25.581395      1.0  \n",
       "2268         25.581395      1.0  \n",
       "2301         25.581395      1.0  \n",
       "2505        100.000000      1.0  \n",
       "...                ...      ...  \n",
       "8043         25.714286      1.0  \n",
       "8059         25.714286      1.0  \n",
       "8433        100.000000      1.0  \n",
       "8476          0.000000      2.0  \n",
       "8480          0.000000      2.0  \n",
       "8485          0.000000      2.0  \n",
       "8491          0.000000      2.0  \n",
       "8495          0.000000      2.0  \n",
       "8499          0.000000      2.0  \n",
       "8504          0.000000      2.0  \n",
       "8510          0.000000      2.0  \n",
       "8515          0.000000      2.0  \n",
       "8521          0.000000      2.0  \n",
       "8527          0.000000      2.0  \n",
       "8533          0.000000      2.0  \n",
       "8538          0.000000      2.0  \n",
       "8543          0.000000      2.0  \n",
       "8548          0.000000      2.0  \n",
       "8553          0.000000      2.0  \n",
       "8557         25.000000      1.0  \n",
       "8559         25.000000      1.0  \n",
       "8562         25.000000      1.0  \n",
       "8565         25.000000      1.0  \n",
       "8568         12.500000      1.0  \n",
       "8578         12.500000      1.0  \n",
       "8587         12.500000      1.0  \n",
       "8597         12.500000      1.0  \n",
       "8606         12.500000      1.0  \n",
       "8614         12.500000      1.0  \n",
       "8622         12.500000      1.0  \n",
       "\n",
       "[200 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rank transcripts by percent recovery\n",
    "txdf['tx_rank'] = txdf.groupby(\"gene\")[\"tx_perc_recovery\"].rank(\"dense\", ascending=False)\n",
    "txdf.head(200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "txdf_fileout = today + '_refseq_VS_clinvar_TX_GENE_RECOVERY.tsv'\n",
    "txdf.to_csv(txdf_fileout, sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# did file save succesfully?\n",
    "os.path.isfile(txdf_fileout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_txdf = txdf[txdf.tx_rank == 1.0]\n",
    "top_txdf = top_txdf[['gene', 'tx_id', 'gene_recovery', 'transcript_recovery', 'tx_perc_recovery', 'tx_rank']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>tx_id</th>\n",
       "      <th>gene_recovery</th>\n",
       "      <th>transcript_recovery</th>\n",
       "      <th>tx_perc_recovery</th>\n",
       "      <th>tx_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>ISG15</td>\n",
       "      <td>NM_005101</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>AGRN</td>\n",
       "      <td>NM_001305275</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>AGRN</td>\n",
       "      <td>NM_198576</td>\n",
       "      <td>30.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>345</th>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>NM_003327</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>366</th>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>NM_080605</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene         tx_id  gene_recovery  transcript_recovery  \\\n",
       "194    ISG15     NM_005101            3.0                  3.0   \n",
       "196     AGRN  NM_001305275           30.0                 15.0   \n",
       "235     AGRN     NM_198576           30.0                 15.0   \n",
       "345  TNFRSF4     NM_003327            1.0                  1.0   \n",
       "366  B3GALT6     NM_080605           18.0                 18.0   \n",
       "\n",
       "     tx_perc_recovery  tx_rank  \n",
       "194             100.0      1.0  \n",
       "196              50.0      1.0  \n",
       "235              50.0      1.0  \n",
       "345             100.0      1.0  \n",
       "366             100.0      1.0  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_txdf.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_txdf_fileout = today + '_refseq_VS_clinvar_TOP_TX_BY_RECOVERY.tsv'\n",
    "top_txdf.to_csv(top_txdf_fileout, sep='\\t', header=True, index=False)\n",
    "\n",
    "# did file save succesfully?\n",
    "os.path.isfile(top_txdf_fileout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Users/pbousounis/Experiments/2019-10-29_hg19mod/2019-10-25_RefSeq_GFF3_download_validate'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "txdf.to_csv('2019-11-06_RefSeq_Ranked_TXs.tsv', sep='\\t', header=True, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
